File: pkg/planner/core/rule/rule_outer_join_to_semi_join.go
// Copyright 2025 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package rule

import (
	"context"

	"github.com/pingcap/tidb/pkg/expression"
	"github.com/pingcap/tidb/pkg/parser/ast"
	"github.com/pingcap/tidb/pkg/parser/mysql"
	"github.com/pingcap/tidb/pkg/planner/core/base"
	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
	"github.com/pingcap/tidb/pkg/util/intset"
)

// OuterJoinToSemiJoin is a logical optimization rule that rewrites a `LogicalJoin` (Outer Join)
// into an `AntiSemiJoin`. The transformation is triggered by a subsequent `LogicalSelection`
// that effectively filters for non-matched rows from the outer join.
//
// The core idea is to identify queries whose semantics are to find rows from one table that
// have NO match in another. This rule recognizes such patterns and changes the `LogicalJoin`'s
// `JoinType` to `AntiSemiJoin`.
//
// A key part of this transformation is the creation of a `LogicalProjection` on top of the
// new `AntiSemiJoin`. This projection is responsible for generating the `NULL` values for the
// columns of the inner table, which is the expected result for this type of query.
//
// The rule is triggered if the `WHERE` clause checks for `IS NULL` on a column from the
// inner table that is guaranteed to be non-null if a match had occurred. This guarantee
// comes from two main patterns identified in the `canConvertAntiJoin` function:
//
//  1. The `IS NULL` check is on a column that is part of the join condition.
//     SQL Example: `SELECT B.* FROM A RIGHT JOIN B ON A.id = B.a_id WHERE A.id IS NULL`
//
//  2. The `IS NULL` check is on a column that has a `NOT NULL` constraint in its table definition.
//     SQL Example: `SELECT A.* FROM A LEFT JOIN B ON A.id = B.a_id WHERE B.non_null_col IS NULL`
//
// In both cases, the original `LogicalSelection` is eliminated, and the plan is rewritten to
// `LogicalProjection -> LogicalJoin(AntiSemiJoin)`.
type OuterJoinToSemiJoin struct{}

// Optimize implements base.LogicalOptRule.<0th> interface.
func (o *OuterJoinToSemiJoin) Optimize(_ context.Context, p base.LogicalPlan) (base.LogicalPlan, bool, error) {
	result, isChanged := o.recursivePlan(p)
	return result, isChanged, nil
}

func (o *OuterJoinToSemiJoin) recursivePlan(p base.LogicalPlan) (base.LogicalPlan, bool) {
	var isChanged bool
	if sel, ok := p.(*logicalop.LogicalSelection); ok {
		p, changed := o.dealWithSelection(nil, 0, sel)
		isChanged = isChanged || changed
		return p, isChanged
	}
	for idx, child := range p.Children() {
		var changed bool
		sel, ok := child.(*logicalop.LogicalSelection)
		if !ok {
			_, changed = o.recursivePlan(child)
			isChanged = isChanged || changed
			continue
		}
		_, changed = o.dealWithSelection(p, idx, sel)
		isChanged = isChanged || changed
	}
	return p, isChanged
}

func (o *OuterJoinToSemiJoin) dealWithSelection(p base.LogicalPlan, childIdx int, sel *logicalop.LogicalSelection) (base.LogicalPlan, bool) {
	selChild := sel.Children()[0]
	switch cc := selChild.(type) {
	case *logicalop.LogicalJoin:
		newP, changed := o.startConvertOuterJoinToSemiJoin(p, childIdx, sel, cc)
		return ensureSelectionRoot(newP, sel), changed
	case *logicalop.LogicalProjection:
		if validProjForConvertAntiJoin(cc) {
			projectionChild := cc.Children()[0]
			join, ok := projectionChild.(*logicalop.LogicalJoin)
			if ok {
				newP, changed := o.startConvertOuterJoinToSemiJoin(p, childIdx, sel, join)
				return ensureSelectionRoot(newP, sel), changed
			}
			newChild, changed := o.recursivePlan(projectionChild)
			resetChildIfChanged(cc, projectionChild, newChild)
			return ensureSelectionRoot(p, sel), changed
		}
		newChild, changed := o.recursivePlan(cc)
		resetChildIfChanged(sel, cc, newChild)
		return ensureSelectionRoot(p, sel), changed
	}
	newChild, changed := o.recursivePlan(selChild)
	resetChildIfChanged(sel, selChild, newChild)
	return ensureSelectionRoot(p, sel), changed
}

func (o *OuterJoinToSemiJoin) startConvertOuterJoinToSemiJoin(p base.LogicalPlan, childIdx int, sel *logicalop.LogicalSelection, join *logicalop.LogicalJoin) (base.LogicalPlan, bool) {
	proj, ok := canConvertAntiJoin(join, sel.Conditions, sel.Schema())
	if ok {
		var pChild base.LogicalPlan
		if proj != nil {
			// projection <- anti semi join
			proj.SetChildren(join)
			pChild = proj
		} else {
			// anti semi join only
			pChild = join
		}
		if p != nil {
			p.SetChild(childIdx, pChild)
		} else {
			p = pChild
		}
	}
	_, changed := o.recursivePlan(join)
	return p, ok || changed
}

func ensureSelectionRoot(p base.LogicalPlan, sel *logicalop.LogicalSelection) base.LogicalPlan {
	if p == nil {
		return sel
	}
	return p
}

func resetChildIfChanged(parent, oldChild, newChild base.LogicalPlan) {
	if newChild != oldChild {
		parent.SetChildren(newChild)
	}
}

// Name implements base.LogicalOptRule.<1st> interface.
func (*OuterJoinToSemiJoin) Name() string {
	return "outer_join_to_semi_join"
}

// canConvertAntiJoin is used in outer-join-to-semi-join rule.
/*
#### Scenario 1: IS NULL on the Join Condition Column

- In this scenario, the IS NULL filter is applied directly to the join key from the inner table.

##### Table Schema
CREATE TABLE Table_A (id INT, name VARCHAR(50));
CREATE TABLE Table_B (id INT, info VARCHAR(50));

##### Plan
```
-- The optimizer rewrites the LEFT JOIN ... WHERE ... IS NULL pattern into an efficient
-- Anti Semi Join to retrieve rows from Table A that have no match in Table B.
SELECT Table_A.id, Table_A.name
FROM
	Table_A
LEFT JOIN
	Table_B ON Table_A.id = Table_B.id
WHERE
	Table_B.id IS NULL;

=>

SELECT Table_A.id, Table_A.name
FROM
	Table_A
WHERE NOT EXISTS (
	SELECT 1
	FROM
		Table_B
	WHERE
		Table_A.id = Table_B.id
);
```
#### Scenario 2: IS NULL on a Non-Join NOT NULL Column

If a column in the inner table is defined as `NOT NULL` in the schema but is filtered as `IS NULL` after the join,
and that column comes from the null-supplied side of the outer join, this indicates that the join did not find a match.
This allows us to convert the join into an `ANTI SEMI JOIN`, even if the column is not part of the join keys.

##### Table Schema:

CREATE TABLE Table_A (id INT,name VARCHAR(50));
CREATE TABLE Table_B (
	id INT,
	status VARCHAR(20) NOT NULL -- Non-join column with NOT NULL constraint
);

##### Plan
```
-- Even though 'status' is not a join key, its NOT NULL constraint
-- ensures that 'B.status IS NULL' only occurs when no match is found.
SELECT A.* FROM Table_A A
LEFT JOIN Table_B B ON A.id = B.id
WHERE B.status IS NULL;

=>
SELECT A.* FROM Table_A A
WHERE NOT EXISTS (
	SELECT 1 FROM Table_B B WHERE A.id = B.id
);
```

Additionally, we found that there may be a projection between select and join.
	Select[isnull(col1)] <- Projection[col1,col2,col3,col4] <- left outer join[outer side: col1 col2, inner side: col3 col4]
Currently, we only support projections with column mapping relationships, not transformation relationships.
	Projection[null->col1,null->col2,col3,col4] <- anti semi join[outer side: col1 col2, inner side: col3 col4]
*/
func canConvertAntiJoin(p *logicalop.LogicalJoin, selectCond []expression.Expression, selectSch *expression.Schema) (resultProj *logicalop.LogicalProjection, canConvertToAntiSemiJoin bool) {
	if len(selectCond) != 1 || (len(p.EqualConditions) == 0 && len(p.OtherConditions) == 0) {
		// This optimization only supports a single selection condition in selectCond.
		// Any condition that refers only to the outer side of the join can be pushed below the join,
		// so the remaining selection condition must be on the inner side.
		// If multiple inner-side conditions were semantically equivalent, the optimizer should keep just one
		// and eliminate the others, so we require exactly one selection condition here.
		return nil, false
	}
	if _, ok := p.Self().(*logicalop.LogicalApply); ok {
		// For now, we won't add complexity in the outer-join-to-semi-join rule.
		// Additionally, existing user cases haven't been applied yet; they can be added later if needed.
		return nil, false
	}
	var outerChildIdx int
	switch p.JoinType {
	case base.LeftOuterJoin:
		outerChildIdx = 0
	case base.RightOuterJoin:
		outerChildIdx = 1
	default:
		return nil, false
	}

	sf, ok := selectCond[0].(*expression.ScalarFunction)
	if !ok {
		return nil, false
	}
	if sf.FuncName.L != ast.IsNull {
		return nil, false
	}
	args := sf.GetArgs()
	// Get the Column in the IsNull
	isNullCol, ok := args[0].(*expression.Column)
	if !ok {
		return nil, false
	}
	outer := p.Children()[outerChildIdx]
	outerSchema := outer.Schema()
	innerSchemaSet := intset.NewFastIntSet()
	// Identify all columns from the inner side of the join (columns that are not in the outer schema).
	// If the column referenced in the IS NULL predicate is one of these inner-side columns used in
	// equal/other join conditions, the outer join can be safely rewritten as an anti-semi join.
	expression.ExtractColumnsSetFromExpressions(&innerSchemaSet, func(c *expression.Column) bool {
		return !outerSchema.Contains(c)
	}, expression.Column2Exprs(p.Schema().Columns)...)
	// Scenario 1: column in IsNull expression is from the inner side columns in the eq/other condition.
	joinCondNRInnerCol := joinCondNullRejectsInnerCol(&innerSchemaSet, isNullCol.UniqueID, p.EqualConditions, p.OtherConditions)

	// Scenario 2:
	//  column in IsNull expression is from the inner side columns.
	//  but it is not in the equal/other condition.
	//  We need to check whether inner column is not null in the origin table schema.
	//  If it is not null column, it can be directly converted into an anti-semi join.
	innerSch := p.Children()[1^outerChildIdx].Schema()
	// if isNullColInnerSchIdx < 0, it means the is null column is not in the inner schema.
	isNullColInnerSchIdx := innerSch.ColumnIndex(isNullCol)
	isInnerNonNullCol := isNullColInnerSchIdx >= 0 &&
		mysql.HasNotNullFlag(innerSch.Columns[isNullColInnerSchIdx].RetType.GetFlag())

	// If either scenario is met, it can be converted to an anti-semi join.
	canConvertToAntiSemiJoin = joinCondNRInnerCol || isInnerNonNullCol
	if canConvertToAntiSemiJoin {
		// resultProj is to generate the NULL values for the columns of the inner table, which is the
		// expected result for this kind of anti-join query.
		resultProj = generateProjectForConvertAntiJoin(p, &innerSchemaSet, outerSchema, selectSch)
		// Anti-semi join's first child is outer, the second child is inner. join condition is outer op inner
		// right outer join's first child is inner, the second child is outer, join condition is inner op outer
		// left outer join's  first child is outer, the second child is inner. join condition is outer op inner
		// So we have to swap it here.
		ctx := p.SCtx().GetExprCtx()
		if p.JoinType == base.RightOuterJoin {
			for idx, expr := range p.EqualConditions {
				args := expr.GetArgs()
				p.EqualConditions[idx] = expression.NewFunctionInternal(ctx, expr.FuncName.L, expr.GetType(ctx.GetEvalCtx()), args[1], args[0]).(*expression.ScalarFunction)
			}
			// it is not needed for the other conditions because the arguments order does not matter.
			args := p.Children()
			p.SetChildren(args[1], args[0])
			tmp := p.LeftConditions
			p.LeftConditions = p.RightConditions
			p.RightConditions = tmp
		}
		p.JoinType = base.AntiSemiJoin
		p.MergeSchema()
	}
	return resultProj, canConvertToAntiSemiJoin
}

func validProjForConvertAntiJoin(proj *logicalop.LogicalProjection) bool {
	if proj == nil {
		return true
	}
	// Sometimes there is a projection between select and join.
	// We require that this projection does not make additional changes to any columns,
	// then we can continue converting to a semi join.
	//  Pass: Projection[col1,col2]
	//  Fail: Projection[col2->col2_renamed, col2+1->col2_incremented]
	for idx, c := range proj.Schema().Columns {
		if !c.Equals(proj.Exprs[idx]) {
			return false
		}
	}
	return true
}

// joinCondNullRejectsInnerCol checks whether the IS NULL column (above the join) is from the inner
// side and appears in the join predicate that null-rejects it (excluding NullEQ; only GT/GE/LE/LT/NE in others).
func joinCondNullRejectsInnerCol(innerSchSet *intset.FastIntSet, isNullColumnID int64, eq []*expression.ScalarFunction, other []expression.Expression) bool {
	for _, s := range eq {
		if s.FuncName.L == ast.NullEQ {
			continue
		}
		col1, col2, ok := expression.IsColOpCol(s)
		if ok {
			if innerSchSet.Has(int(col1.UniqueID)) {
				if isNullColumnID == col1.UniqueID {
					return true
				}
			} else if innerSchSet.Has(int(col2.UniqueID)) {
				if isNullColumnID == col2.UniqueID {
					return true
				}
			}
		}
	}
	for _, s := range other {
		if sf, ok := s.(*expression.ScalarFunction); ok {
			switch sf.FuncName.L {
			case ast.GT, ast.GE, ast.LE, ast.LT, ast.NE:
				col1, col2, ok := expression.IsColOpCol(sf)
				if ok {
					if innerSchSet.Has(int(col1.UniqueID)) && isNullColumnID == col1.UniqueID {
						return true
					} else if innerSchSet.Has(int(col2.UniqueID)) && isNullColumnID == col2.UniqueID {
						return true
					}
				}
			default:
			}
		}
	}
	return false
}

// generateProjectForConvertAntiJoin is to generate projection and put it on the anti-semi join which is from outer join.
// outer column will not be changed. inner column will output the null.
func generateProjectForConvertAntiJoin(p *logicalop.LogicalJoin, innerSchSet *intset.FastIntSet, outerSchema, parentNodeSchema *expression.Schema) (proj *logicalop.LogicalProjection) {
	isAllFromOuterSide := true
	for _, c := range parentNodeSchema.Columns {
		if !outerSchema.Contains(c) {
			isAllFromOuterSide = false
			break
		}
	}
	if isAllFromOuterSide {
		return nil
	}
	projExprs := make([]expression.Expression, 0, len(parentNodeSchema.Columns))
	for _, c := range parentNodeSchema.Columns {
		if innerSchSet.Has(int(c.UniqueID)) {
			projExprs = append(projExprs, expression.NewNull())
		} else {
			projExprs = append(projExprs, c.Clone())
		}
	}
	proj = logicalop.LogicalProjection{Exprs: projExprs}.Init(p.SCtx(), p.QueryBlockOffset())
	proj.SetSchema(parentNodeSchema.Clone())
	return
}

File: rust/crates/tidb-planner/src/logical/rule_outer_join_to_semi_join.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Go `pkg/planner/core/rule/rule_outer_join_to_semi_join.go`.

use std::collections::BTreeSet;

use tidb_datatype::{Datum, FieldTypeFlags};
use tidb_expr::constant::Constant;
use tidb_expr::expr_util::extract::is_col_op_col;
use tidb_expr::expression::Expression;
use tidb_expr::schema::Schema;

use crate::find_best_task::LogicalJoinType;
use crate::logical::{BaseLogicalPlan, LogicalJoin, LogicalPlan, LogicalProjection};
use crate::plan_base::PlanError;

use super::rule::{LogicalOptRule, RuleContext};

/// Go `OuterJoinToSemiJoin`.
#[derive(Debug)]
pub struct OuterJoinToSemiJoin;

impl LogicalOptRule for OuterJoinToSemiJoin {
    #[allow(clippy::result_large_err)]
    fn optimize(
        &self,
        ctx: &RuleContext<'_>,
        plan: LogicalPlan,
    ) -> Result<(LogicalPlan, bool), (LogicalPlan, PlanError)> {
        match rewrite(ctx, plan) {
            Ok(result) => Ok(result),
            Err((plan, error)) => Err((plan, error)),
        }
    }

    fn name(&self) -> &'static str {
        "outer_join_to_semi_join"
    }
}

#[allow(clippy::result_large_err)]
fn rewrite(
    ctx: &RuleContext<'_>,
    mut plan: LogicalPlan,
) -> Result<(LogicalPlan, bool), (LogicalPlan, PlanError)> {
    if matches!(plan, LogicalPlan::Selection(_)) {
        return rewrite_selection(ctx, plan);
    }
    let children = std::mem::take(plan.base_mut().children_mut());
    let mut rewritten = Vec::with_capacity(children.len());
    let mut changed = false;
    for child in children {
        match rewrite(ctx, child) {
            Ok((child, child_changed)) => {
                rewritten.push(child);
                changed |= child_changed;
            }
            Err((child, error)) => {
                rewritten.push(child);
                plan.set_children(rewritten);
                return Err((plan, error));
            }
        }
    }
    plan.set_children(rewritten);
    Ok((plan, changed))
}

#[allow(clippy::result_large_err)]
fn rewrite_selection(
    ctx: &RuleContext<'_>,
    plan: LogicalPlan,
) -> Result<(LogicalPlan, bool), (LogicalPlan, PlanError)> {
    let LogicalPlan::Selection(mut selection) = plan else {
        unreachable!()
    };
    let parent_schema = selection.base.base.schema().cloned().or_else(|| {
        selection
            .base
            .children()
            .first()
            .and_then(LogicalPlan::schema)
            .cloned()
    });
    let parent_names = if selection.base.base.output_names().is_empty() {
        selection
            .base
            .children()
            .first()
            .map_or_else(Vec::new, |child| child.output_names().to_vec())
    } else {
        selection.base.base.output_names().to_vec()
    };
    let Some(child) = selection.base.children_mut().pop() else {
        return Ok((LogicalPlan::Selection(selection), false));
    };

    match child {
        LogicalPlan::Join(join) => {
            match convert(
                ctx,
                join.clone(),
                &selection.conditions,
                parent_schema.as_ref(),
                parent_names,
            ) {
                Ok(Some(converted)) => {
                    rewrite(ctx, converted).map(|(plan, changed)| (plan, true | changed))
                }
                Ok(None) => {
                    selection.base.set_children(vec![LogicalPlan::Join(join)]);
                    let child = selection.base.children_mut().pop().unwrap();
                    match rewrite(ctx, child) {
                        Ok((child, changed)) => {
                            selection.base.set_children(vec![child]);
                            Ok((LogicalPlan::Selection(selection), changed))
                        }
                        Err((child, error)) => {
                            selection.base.set_children(vec![child]);
                            Err((LogicalPlan::Selection(selection), error))
                        }
                    }
                }
                Err(error) => {
                    selection.base.set_children(vec![LogicalPlan::Join(join)]);
                    Err((LogicalPlan::Selection(selection), error))
                }
            }
        }
        LogicalPlan::Projection(mut projection) if valid_projection(&projection) => {
            let Some(projection_child) = projection.base.children_mut().pop() else {
                selection
                    .base
                    .set_children(vec![LogicalPlan::Projection(projection)]);
                return Ok((LogicalPlan::Selection(selection), false));
            };
            if let LogicalPlan::Join(join) = projection_child {
                match convert(
                    ctx,
                    join.clone(),
                    &selection.conditions,
                    parent_schema.as_ref(),
                    parent_names,
                ) {
                    Ok(Some(converted)) => {
                        rewrite(ctx, converted).map(|(plan, changed)| (plan, true | changed))
                    }
                    Ok(None) => {
                        projection.base.set_children(vec![LogicalPlan::Join(join)]);
                        selection
                            .base
                            .set_children(vec![LogicalPlan::Projection(projection)]);
                        let child = selection.base.children_mut().pop().unwrap();
                        match rewrite(ctx, child) {
                            Ok((child, changed)) => {
                                selection.base.set_children(vec![child]);
                                Ok((LogicalPlan::Selection(selection), changed))
                            }
                            Err((child, error)) => {
                                selection.base.set_children(vec![child]);
                                Err((LogicalPlan::Selection(selection), error))
                            }
                        }
                    }
                    Err(error) => {
                        projection.base.set_children(vec![LogicalPlan::Join(join)]);
                        selection
                            .base
                            .set_children(vec![LogicalPlan::Projection(projection)]);
                        Err((LogicalPlan::Selection(selection), error))
                    }
                }
            } else {
                projection.base.set_children(vec![projection_child]);
                match rewrite(ctx, LogicalPlan::Projection(projection)) {
                    Ok((child, changed)) => {
                        selection.base.set_children(vec![child]);
                        Ok((LogicalPlan::Selection(selection), changed))
                    }
                    Err((child, error)) => {
                        selection.base.set_children(vec![child]);
                        Err((LogicalPlan::Selection(selection), error))
                    }
                }
            }
        }
        child => match rewrite(ctx, child) {
            Ok((child, changed)) => {
                selection.base.set_children(vec![child]);
                Ok((LogicalPlan::Selection(selection), changed))
            }
            Err((child, error)) => {
                selection.base.set_children(vec![child]);
                Err((LogicalPlan::Selection(selection), error))
            }
        },
    }
}

fn valid_projection(projection: &LogicalProjection) -> bool {
    let Some(schema) = projection.base.base.schema() else {
        return false;
    };
    schema.columns.len() == projection.exprs.len()
        && schema
            .columns
            .iter()
            .zip(&projection.exprs)
            .all(|(column, expression)| Expression::Column(column.clone()).equal(expression))
}

fn join_condition_null_rejects(
    join: &LogicalJoin,
    inner_ids: &BTreeSet<i64>,
    is_null_column: i64,
) -> bool {
    let rejects = |expression: &Expression| {
        let Expression::ScalarFunction(function) = expression else {
            return false;
        };
        let Some((left, right)) = is_col_op_col(function) else {
            return false;
        };
        (inner_ids.contains(&left.unique_id) && left.unique_id == is_null_column)
            || (inner_ids.contains(&right.unique_id) && right.unique_id == is_null_column)
    };
    join.equal_conditions
        .iter()
        .filter(|condition| condition.func_name.lowercase() != "nulleq")
        .any(|condition| rejects(&Expression::ScalarFunction(condition.clone())))
        || join.other_conditions.iter().any(|condition| {
            matches!(condition, Expression::ScalarFunction(function)
                if matches!(function.func_name.lowercase(), "gt" | "ge" | "le" | "lt" | "ne"))
                && rejects(condition)
        })
}

#[allow(clippy::result_large_err)]
fn convert(
    ctx: &RuleContext<'_>,
    mut join: LogicalJoin,
    selection_conditions: &[Expression],
    parent_schema: Option<&Schema>,
    parent_names: Vec<tidb_datatype::FieldName>,
) -> Result<Option<LogicalPlan>, PlanError> {
    if selection_conditions.len() != 1
        || (join.equal_conditions.is_empty() && join.other_conditions.is_empty())
    {
        return Ok(None);
    }
    let outer_index = match join.join_type {
        LogicalJoinType::LeftOuter => 0,
        LogicalJoinType::RightOuter => 1,
        _ => return Ok(None),
    };
    let Expression::ScalarFunction(is_null) = &selection_conditions[0] else {
        return Ok(None);
    };
    let [Expression::Column(is_null_column)] = is_null.get_args() else {
        return Ok(None);
    };
    if is_null.func_name.lowercase() != "isnull" {
        return Ok(None);
    }
    let Some(outer_schema) = join
        .base
        .children()
        .get(outer_index)
        .and_then(LogicalPlan::schema)
        .cloned()
    else {
        return Ok(None);
    };
    let inner_index = 1 ^ outer_index;
    let Some(inner_schema) = join
        .base
        .children()
        .get(inner_index)
        .and_then(LogicalPlan::schema)
        .cloned()
    else {
        return Ok(None);
    };
    let inner_ids = inner_schema
        .columns
        .iter()
        .map(|column| column.unique_id)
        .collect::<BTreeSet<_>>();
    let inner_non_null = inner_schema
        .retrieve_column(is_null_column)
        .and_then(|column| column.get_static_type())
        .is_some_and(|field_type| field_type.has_flag(FieldTypeFlags::NOT_NULL));
    if !join_condition_null_rejects(&join, &inner_ids, is_null_column.unique_id) && !inner_non_null
    {
        return Ok(None);
    }

    if join.join_type == LogicalJoinType::RightOuter {
        for condition in &mut join.equal_conditions {
            let args = condition.args.clone();
            let expression = ctx
                .builder
                .new_function(
                    condition.func_name.lowercase(),
                    condition.ret_type.clone(),
                    vec![args[1].clone(), args[0].clone()],
                )
                .map_err(|error| PlanError::internal(error.to_string()))?;
            let Expression::ScalarFunction(rebuilt) = expression else {
                return Err(PlanError::internal(
                    "join equality did not rebuild as a scalar function",
                ));
            };
            *condition = rebuilt;
        }
        join.base.children_mut().swap(0, 1);
        std::mem::swap(&mut join.left_conditions, &mut join.right_conditions);
    }
    join.join_type = LogicalJoinType::AntiSemi;
    join.base.base.set_schema(Some(outer_schema.clone()));
    let outer_names = join
        .base
        .children()
        .first()
        .map_or_else(Vec::new, |child| child.output_names().to_vec());
    join.base.base.set_output_names(outer_names);
    let join = LogicalPlan::Join(join);

    let Some(parent_schema) = parent_schema else {
        return Ok(Some(join));
    };
    if parent_schema
        .columns
        .iter()
        .all(|column| outer_schema.contains(column))
    {
        return Ok(Some(join));
    }
    let expressions = parent_schema
        .columns
        .iter()
        .map(|column| {
            if inner_ids.contains(&column.unique_id) {
                let mut field_type = column.get_static_type().cloned().unwrap_or_else(|| {
                    tidb_datatype::FieldType::new(tidb_datatype::FieldTypeCode::Tiny)
                });
                field_type.del_flags(FieldTypeFlags::NOT_NULL);
                Expression::Constant(Constant::new(Datum::Null, field_type))
            } else {
                Expression::Column(column.clone())
            }
        })
        .collect();
    let mut base = BaseLogicalPlan::new(
        ctx.allocator,
        LogicalProjection::TYPE,
        join.base().base.query_block_offset(),
    );
    base.base.set_schema(Some(parent_schema.clone()));
    base.base.set_output_names(parent_names);
    let mut projection = LogicalPlan::Projection(LogicalProjection::new(base, expressions));
    projection.set_children(vec![join]);
    Ok(Some(projection))
}
